File: ruby/solver.rb
def find_cell(grid, x, y)
	for j in y..8
		for i in x..8
			if grid[i][j] == nil
				return false, i, j
			end
		end
		x = 0
	end

	return true, -1, -1, nil
end

def solve(grid, x = 0, y = 0)
	is_finished, x, y = find_cell(grid, x, y)

	return true if is_finished

	col = grid.map { |l| l[y] }.compact
	line = grid[x].compact

	for i in 1..9

		if col.include?(i) || line.include?(i)
			next
		end

		grid[x][y] = i

		if !check(grid)
			next
		end

		if (x == 8)
			if solve(grid, 0, y + 1)
				return true
			end
		else 
			if solve(grid, x + 1, y)
				return true
			end
		end
	end

	# we are going to backtrack
	grid[x][y] = nil

	return false
end

def check(grid)
	# Here we check if a line or column have the same element twice
	for i in 0..8
		col = grid.map { |l| l[i] }
		line = grid[i]

		# compact remove the nil
		# but we don't use compact! because xe don't want to alter the initial grid
		col = col.compact
		if col.size != col.uniq.size
			return false
		end

		line = line.compact
		if line.compact.size != line.uniq.size
			return false
		end
	end

	for y in 0..2
		for x in 0..2
			block = []
			for index in 0..2
				block += grid[(x * 3) + index][(y * 3)..((y + 1) * 3 - 1)]
			end

			block = block.compact
			if block.size != block.uniq.size
				return false
			end
		end
	end

	return true
end

File: go/src/dumb/solver.go
package dumb

import (
	"sudoku"
)

var SIZE_OF_SUDOKU byte = sudoku.SIZE_OF_SUDOKU
var UdefVal byte = sudoku.UdefVal

func find_cell(grid []byte, xp, yp byte) (finished bool, x, y byte) {
	x, y = xp, yp
	for ; y < SIZE_OF_SUDOKU; y++ {
		for ; x < SIZE_OF_SUDOKU; x++ {
			if grid[sudoku.HASH(x, y)] == UdefVal {
				finished = false
				return
			}
		}
		x = 0
	}

	finished = true
	return
}

func include(grid []byte, x, y, value byte) bool {
	for n := byte(0); n < SIZE_OF_SUDOKU; n++ {
		switch {
		case grid[sudoku.HASH(n, y)] == value:
			return true
		case grid[sudoku.HASH(x, n)] == value:
			return true
		case grid[sudoku.HASH_BLOC(x, y, n)] == value:
			return true
		}
	}
	return false
}

func Solve(grid []byte, x, y byte) bool {
	var finished bool

	if finished, x, y = find_cell(grid, x, y); finished {
		return true
	}

	for i := byte(1); i <= SIZE_OF_SUDOKU; i++ {
		if include(grid, x, y, i) {
			continue
		}

		grid[sudoku.HASH(x, y)] = i

		if x == (SIZE_OF_SUDOKU - 1) {
			if Solve(grid, 0, y + 1) {
				return true
			}
		} else {
			if Solve(grid, x + 1, y) {
				return true
			}
		}
	}

	// we are going to backtrack
	grid[sudoku.HASH(x, y)] = UdefVal

	return false

}

